### PR TITLE
lint using pre-commit

### PR DESCRIPTION
**Pull request by danieleades**
here's another experiment in automatic linting, this time using [pre-commit](https://pre-commit.com/)

i've also added

- flake8-bugbear: opinionated lints
- pyupgrade: checks for, and automatically updates deprecated syntax

### CODE REVIEW

File: pyproject.toml
Comment (danieleades) [inline pyproject.toml:32]: this is a pre-commit requirement. it's possible to avoid bumping this version, by instead specifying that pre-commit is installed as a system dependency, instead of using Poetry. do you have a preference?

File: docs/contributing.rst
Comment (danwos) [inline docs/contributing.rst:86]: This is markdown syntax and does not get rendered. :)
Please stay with rst.

File: docs/contributing.rst
Comment (danwos) [inline docs/contributing.rst:95]: Not so happy with the docs.
I think pre-commit gets not introduced to support linting.
It's more or less a general quality gate which needs to be passed before a commit can be pushed.

So we need paragraph how to create and send PRs to this project.
If you like, you can already add it.
Or I will do it with a new PR (after yours got merged).

File: pyproject.toml
Comment (danwos) [reply pyproject.toml:32]: It's fine to have it as project dependency, as the version diff is not really soo huge :)

**Review by danwos — CHANGES_REQUESTED** [review]

File: docs/contributing.rst
Comment (danieleades) [reply docs/contributing.rst]: my bad! I live in a world of markdown usually.

I *think* i've fixed this properly, but please give this a sanity-check for me!

File: docs/contributing.rst
Comment (danieleades) [reply docs/contributing.rst:96]: can you suggest a wording that you prefer?

or i can leave it to you to update in a following commit? whichever you prefer

**Review by danwos — APPROVED** [review]

**Comment by danieleades** [comment]
this PR was a little ambitious anyway- too much in one go.

closing in favour of #415

**Pull request closed without merging**
